### PR TITLE
Add null annotations to registries

### PR DESCRIPTION
**Pull request by wborn**
Adds null annotations to all registries, the interfaces they implement and a few other classes.

---

There's also a PR to update the add-ons for this: https://github.com/openhab/openhab2-addons/pull/6233

### CODE REVIEW

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/module/provider/i18n/ModuleTypeI18nService.java [hunk 1]
```diff
@@ -32,6 +35,8 @@
      * @param bundle the bundle containing the localization files
      * @return the localized ModuleType
      */
-    ModuleType getModuleTypePerLocale(ModuleType defModuleType, Locale locale, Bundle bundle);
+    @Nullable
+    ModuleType getModuleTypePerLocale(@Nullable ModuleType defModuleType, @Nullable Locale locale,
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/module/provider/i18n/ModuleTypeI18nService.java]: Are you sure the `Bundle` should be annotated with `@Nullable`? Because if we do not have a bundle we will not get a translation. Thus I think is does not make sense to call such a method in that case. We e.g. did not annotated it with `@Nullable` in the `ConfigI18nLocalizationService` or `ProfileI18nUtil`.

Comment (wborn) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/module/provider/i18n/ModuleTypeI18nService.java]: Looks like I used `@Nullable` because the `bundle` in `org.openhab.core.automation.internal.provider.ModuleTypeResourceBundleProvider.getPerLocale` can be `null` :

```java
        String uid = defModuleType.getUID();
        Bundle bundle = getBundle(uid);
```

We can also let `getPerLocale` returns `null` when it can't find the bundle.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/i18n/ModuleI18nUtil.java [hunk 2]
```diff
@@ -31,22 +33,28 @@
...
         List<T> lmodules = new ArrayList<>();
         for (T module : modules) {
             String label = getModuleLabel(i18nProvider, bundle, uid, module.getId(), module.getLabel(), prefix, locale);
             String description = getModuleDescription(i18nProvider, bundle, uid, prefix, module.getId(),
                     module.getDescription(), locale);
-            lmodules.add(createLocalizedModule(module, label, description));
+            @Nullable
+            T lmodule = createLocalizedModule(module, label, description);
+            if (lmodule != null) {
+                lmodules.add(lmodule);
```
[the remark this answers is not in the document]
Comment (wborn) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/i18n/ModuleI18nUtil.java]: That might make sense. I sometimes had a hard time to understand what some code is about without any docs and all the generic terminology. :smile:

File: bundles/org.openhab.core.automation.rest/src/main/java/org/openhab/core/automation/rest/internal/ModuleTypeResource.java [hunk 3]
```diff
@@ -62,14 +64,15 @@
  */
 @Path("module-types")
 @Api("module-types")
+@NonNullByDefault
 @Component
 public class ModuleTypeResource implements RESTResource {
 
-    private ModuleTypeRegistry moduleTypeRegistry;
-    private LocaleService localeService;
+    private @NonNullByDefault({}) ModuleTypeRegistry moduleTypeRegistry;
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation.rest/src/main/java/org/openhab/core/automation/rest/internal/ModuleTypeResource.java]: One general thing: I would prefer to use constructor injection instead of this annotation. But I would be fine to do it in a separate PR after this one has been merged.

Comment (wborn) [reply bundles/org.openhab.core.automation.rest/src/main/java/org/openhab/core/automation/rest/internal/ModuleTypeResource.java]: Yes this PR has become too big already so I'd like to wrap it up before I spent the rest of my time on resolving merge conflicts. :wink: 

I hope with having annotated all registries, any subsequent PR that adds null annotations will be easier.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/module/provider/i18n/ModuleTypeI18nService.java (hunk 1 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/module/provider/i18n/ModuleTypeI18nService.java]: I tried to look into this. At the first glance I second your suggestion to return `null` if the `bundle` is `null`.

At the second view I would tend to throw away the `getBundle()` method it somehow looks strange to me. In other bundles we use the `FrameworkUtil.getBundle()` method in such cases. But to be honest I do not now if it will work here.

Comment (wborn) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/module/provider/i18n/ModuleTypeI18nService.java]: Until I fully understand what is going here I've gone for the conservative approach by adding a null check and not throwing the method away. I think the idea is that these ModuleTypes can be provided by different bundles.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/i18n/ModuleI18nUtil.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wborn) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/i18n/ModuleI18nUtil.java]: I think this was fixed by you. :+1:

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/AbstractResourceBundleProvider.java [hunk 4]
```diff
@@ -395,7 +387,8 @@ protected Bundle getBundle(String uid) {
     }
 
     protected List<ConfigDescriptionParameter> getLocalizedConfigurationDescription(TranslationProvider i18nProvider,
-            List<ConfigDescriptionParameter> config, Bundle bundle, String uid, String prefix, Locale locale) {
+            @Nullable List<ConfigDescriptionParameter> config, @Nullable Bundle bundle, String uid, String prefix,
+            Locale locale) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/AbstractResourceBundleProvider.java]:
```suggestion
            @Nullable Locale locale) {
```

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/type/ModuleTypeRegistryImpl.java [hunk 5]
```diff
@@ -73,22 +76,23 @@ public ModuleType get(String typeUID) {
...
         Collection<T> result = new ArrayList<>(20);
         forEach((provider, mType) -> {
             ModuleType mt = locale == null ? mType
                     : ((ModuleTypeProvider) provider).getModuleType(mType.getUID(), locale);
-            Collection<String> tags = mt.getTags();
-            if (moduleTypeTag == null) {
-                result.add((T) createCopy(mt));
-            } else if (tags.contains(moduleTypeTag)) {
-                result.add((T) createCopy(mt));
+            if ((mt != null) && (moduleTypeTag == null || mt.getTags().contains(moduleTypeTag))) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/type/ModuleTypeRegistryImpl.java]:
```suggestion
            if (mt != null && (moduleTypeTag == null || mt.getTags().contains(moduleTypeTag))) {
```

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/type/ModuleTypeRegistryImpl.java [hunk 6]
```diff
@@ -101,23 +105,25 @@ public ModuleType get(String typeUID) {
...
+        Set<String> tagSet = new HashSet<>(Arrays.asList(tags));
         Collection<T> result = new ArrayList<>(20);
         forEach((provider, mType) -> {
             ModuleType mt = locale == null ? mType
                     : ((ModuleTypeProvider) provider).getModuleType(mType.getUID(), locale);
-            if (tagSet == null) {
-                result.add((T) createCopy(mt));
-            } else if (mt.getTags().containsAll(tagSet)) {
-                result.add((T) createCopy(mt));
+            if ((mt != null) && (mt.getTags().containsAll(tagSet))) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/type/ModuleTypeRegistryImpl.java]:
```suggestion
            if (mt != null && (mt.getTags().containsAll(tagSet))) {
```

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/template/TemplateRegistry.java [hunk 7]
```diff
@@ -47,25 +50,25 @@
...
+     *            {@code null} then the result will contain all available templates.
      * @param locale specifies the desired {@link Locale} to be used for localization of the returned elements. If
-     *               localization resources for this locale are not available or the passed locale is {@code null} the
-     *               elements are returned with the default localization.
+     *            localization resources for this locale are not available or the passed locale is {@code null} the
+     *            elements are returned with the default localization.
      * @return a collection of localized templates, which correspond to the specified tag.
      */
-    public Collection<E> getByTag(String tag, Locale locale);
+    public Collection<E> getByTag(@Nullable String tag, Locale locale);
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/template/TemplateRegistry.java]:
```suggestion
    public Collection<E> getByTag(@Nullable String tag, @Nullable Locale locale);
```

File: bundles/org.openhab.core.ui/src/main/java/org/eclipse/smarthome/ui/internal/items/ItemUIRegistryImpl.java [hunk 8]
```diff
@@ -800,7 +804,7 @@ private int getNbOptions(String itemName) {
         return 0;
     }
 
-    private Class<? extends Item> getItemType(@NonNull String itemName) {
+    private @Nullable Class<? extends Item> getItemType(@NonNull String itemName) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.ui/src/main/java/org/eclipse/smarthome/ui/internal/items/ItemUIRegistryImpl.java]:
```suggestion
    private @Nullable Class<? extends Item> getItemType(String itemName) {
```

File: bundles/org.openhab.core.config.core/src/main/java/org/eclipse/smarthome/config/core/i18n/ConfigDescriptionI18nUtil.java [hunk 9]
```diff
@@ -40,29 +43,30 @@ public ConfigDescriptionI18nUtil(TranslationProvider i18nProvider) {
         this.i18nProvider = i18nProvider;
     }
 
-    public String getParameterPattern(Bundle bundle, URI configDescriptionURI, String parameterName,
-            String defaultPattern, Locale locale) {
+    public @Nullable String getParameterPattern(@Nullable Bundle bundle, URI configDescriptionURI, String parameterName,
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/main/java/org/eclipse/smarthome/config/core/i18n/ConfigDescriptionI18nUtil.java]: Same comment like in the `ModuleTypeI18nUtil` regarding the bundle:

```suggestion
    public @Nullable String getParameterPattern(Bundle bundle, URI configDescriptionURI, String parameterName,
```

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/AbstractResourceBundleProvider.java (hunk 4 above)
Comment (wborn) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/AbstractResourceBundleProvider.java]: Many methods got annotated this way but I seemed to have missed that one. :+1:

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/type/ModuleTypeRegistryImpl.java (hunk 5 above)
Comment (wborn) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/type/ModuleTypeRegistryImpl.java]: On other projects my Eclipse is configured to automatically remove redundant parentheses as save cleanup action. :wink:

File: bundles/org.openhab.core.config.core/src/main/java/org/eclipse/smarthome/config/core/i18n/ConfigDescriptionI18nUtil.java (hunk 9 above)
Comment (wborn) [reply bundles/org.openhab.core.config.core/src/main/java/org/eclipse/smarthome/config/core/i18n/ConfigDescriptionI18nUtil.java]: I'll any resolve any null analysis errors this might cause!

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/i18n/ModuleTypeI18nServiceImpl.java [hunk 10]
```diff
@@ -76,35 +86,35 @@ public ModuleType getModuleTypePerLocale(ModuleType defModuleType, Locale locale
...
+            @Nullable Locale locale) {
         try {
-            uri = new URI(prefix + ":" + uid + ".name");
+            return localizationService
+                    .getLocalizedConfigDescription(bundle,
+                            new ConfigDescription(new URI(prefix + ":" + uid + ".name"), parameters), locale)
+                    .getParameters();
         } catch (URISyntaxException e) {
             logger.error("Constructed invalid uri '{}:{}.name'", prefix, uid, e);
+            return null;
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/i18n/ModuleTypeI18nServiceImpl.java]: We should return the default values here. I fixed it in https://github.com/openhab/openhab-core/pull/1161/files#diff-d90b601b2b68bbe8017ba644c98c6da1R387.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/AbstractResourceBundleProvider.java [hunk 11]
```diff
@@ -406,6 +399,7 @@ protected Bundle getBundle(String uid) {
                     uri = new URI(prefix + ":" + uid + ".name");
                 } catch (URISyntaxException e) {
                     logger.error("Constructed invalid uri '{}:{}.name'", prefix, uid, e);
+                    return Collections.emptyList();
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/AbstractResourceBundleProvider.java]: We should return the default values here. I fixed it in https://github.com/openhab/openhab-core/pull/1161/files#diff-d90b601b2b68bbe8017ba644c98c6da1R387.
